### PR TITLE
fix: Refactor package scripts to correctly handle service restart during upgrade

### PR DESCRIPTION
**Pull request by jsirianni**
### Proposed Change

Refactored package scripts to follow a single pattern, which takes the "action" being passed to the script into account.
- pre install
- post install
- pre remove
- post remove

Added inspect tests for package `upgrade` and `remove`.

The four scripts are used during package upgrade (in addition to install / remove), so they need to contain logic which correctly handles this situation. Without considering upgrade scenarios, package upgrades result in the service being stopped during the removal of the previous package. On debian, the service is started again due to the `postinstall` script's debian specific logic. On RHEL based systems, the services is not started (per conversions) and is left with a stopped service without an automatic restart.

The solution is to avoid stopping the service by detecting that we are upgrading, not uninstalling.

Goreleaser has some nice example scripts for detecting install / upgrade / remove. https://nfpm.goreleaser.com/tips/#examples

#### RHEL

Unfortunately, RHEL based systems appear to use the **currently installed** package version's scripts when handling upgrade logic. To test this solution on a RHEL based system, I did the following:

**1. Install v1.12.2 using the install script and ensure the service is running.**
```bash
$ curl -fsSlL https://github.com/observiq/bindplane-op/releases/latest/download/install-linux.sh | bash -s -- --version 1.12.2
```
```
$ sudo systemctl status bindplane
...
   Active: active (running) since Tue 2023-05-02 20:35:27 UTC; 16s ago
...
```
**2. Configure bindplane and restart the service**
```bash
$ sudo /usr/local/bin/bindplane init server --config /etc/bindplane/config.yaml
$ sudo systemctl restart bindplane
```
```
$ sudo systemctl status bindplane
...
   Active: active (running) since Tue 2023-05-02 20:36:29 UTC; 25s ago
...
```
**3. Install a build from this pr using `sudo dnf install <path to rpm file>`**
```bash
$ sudo dnf install bindplane-1.13.0~SNAPSHOT-1a4f0e76.rpm 
```
**4. Observe that the package installs but the service is stopped.**
```
$ sudo systemctl status bindplane
...
   Active: inactive (dead)
...
```
**5. Manualy start bindplane**
```bash
sudo systemctl enable --now bindplane
```
```bash
 sudo systemctl status bindplane
...
   Active: active (running) since Tue 2023-05-02 20:39:32 UTC; 1s ago
...
```
**6. Install a build from this pr (after rebasing, to get a different snapshot build / version)**
```bash
$ sudo dnf install bindplane_1.13.0-SNAPSHOT-f5e9f661_linux_amd64.rpm 
```
**7. Observe that the package installs and the **service** is still running!!**
```
$ sudo systemctl status bindplane
...
   Active: active (running) since Tue 2023-05-02 20:39:32 UTC; 1min 12s ago
...
```

#### Inspec

To test regressions, we can use inspec for install, upgrade, and remove. We expect to see rhel based platforms fail the upgrade tests because the initial package being installed is from v1.12.2, which wills suffer from the issue this PR resolves as explained above.

Inspec output:
```
Instance           Driver                Provisioner  Verifier  Transport  Last Action  Last Error
install-centos-7   Google Compute (GCE)  Shell        Inspec    Ssh        Verified     <None>
install-rocky-8    Google Compute (GCE)  Shell        Inspec    Ssh        Verified     <None>
install-rhel-7     Google Compute (GCE)  Shell        Inspec    Ssh        Verified     <None>
install-rhel-8     Google Compute (GCE)  Shell        Inspec    Ssh        Verified     <None>
install-ubuntu-18  Google Compute (GCE)  Shell        Inspec    Ssh        Verified     <None>
install-ubuntu-20  Google Compute (GCE)  Shell        Inspec    Ssh        Verified     <None>
install-debian-10  Google Compute (GCE)  Shell        Inspec    Ssh        Verified     <None>
install-debian-11  Google Compute (GCE)  Shell        Inspec    Ssh        Verified     <None>
upgrade-centos-7   Google Compute (GCE)  Shell        Inspec    Ssh        Set Up       Kitchen::ActionFailed
upgrade-rocky-8    Google Compute (GCE)  Shell        Inspec    Ssh        Set Up       Kitchen::ActionFailed
upgrade-rhel-7     Google Compute (GCE)  Shell        Inspec    Ssh        Set Up       Kitchen::ActionFailed
upgrade-rhel-8     Google Compute (GCE)  Shell        Inspec    Ssh        Set Up       Kitchen::ActionFailed
upgrade-ubuntu-18  Google Compute (GCE)  Shell        Inspec    Ssh        Verified     <None>
upgrade-ubuntu-20  Google Compute (GCE)  Shell        Inspec    Ssh        Verified     <None>
upgrade-debian-10  Google Compute (GCE)  Shell        Inspec    Ssh        Verified     <None>
upgrade-debian-11  Google Compute (GCE)  Shell        Inspec    Ssh        Verified     <None>
remove-centos-7    Google Compute (GCE)  Shell        Inspec    Ssh        Verified     <None>
remove-rocky-8     Google Compute (GCE)  Shell        Inspec    Ssh        Verified     <None>
remove-rhel-7      Google Compute (GCE)  Shell        Inspec    Ssh        Verified     <None>
remove-rhel-8      Google Compute (GCE)  Shell        Inspec    Ssh        Verified     <None>
remove-ubuntu-18   Google Compute (GCE)  Shell        Inspec    Ssh        Verified     <None>
remove-ubuntu-20   Google Compute (GCE)  Shell        Inspec    Ssh        Verified     <None>
remove-debian-10   Google Compute (GCE)  Shell        Inspec    Ssh        Verified     <None>
remove-debian-11   Google Compute (GCE)  Shell        Inspec    Ssh        Verified     <None>
```

### Known Issues

SLES seems to have trouble with upgrade and uninstall. These issues are not obvious to me, and I have opted to ticket them for a later time, as SLES is not widely adopted by our current user base.

##### Checklist
- [x] Changes are tested
- [x] CI has passed

### CODE REVIEW

File: scripts/package/preinstall.sh [hunk 1]
```diff
@@ -26,17 +26,41 @@
...
-
-if id "$username" &>/dev/null; then
-    echo "User ${username} already exists"
-    exit 0
-else
-    useradd --shell /sbin/nologin --system "$username" -g "$username"
-fi
+# Install creates the bindplane user and group using the
+# name 'bindplane'. The bindplane user does not have a shell.
+# This function can be called as it is idempotent.
```
Comment (andykellr) [inline scripts/package/preinstall.sh:31]:
```suggestion
# This function can be called more than once as it is idempotent.
```

**Review by andykellr — CHANGES_REQUESTED** [review]
There are conflicts in ci.yaml.

I'm not sure the best way to test this. The output of your testing shows that it is still running but doesn't make it clear what version is running. I assume the new one, but I couldn't tell. The timestamp seems to suggest it is the old version still running.

```
(running) since Tue 2023-05-02 20:39:32 UTC
```

**Comment by jsirianni** [comment]
> There are conflicts in ci.yaml.
> 
> I'm not sure the best way to test this. The output of your testing shows that it is still running but doesn't make it clear what version is running. I assume the new one, but I couldn't tell. The timestamp seems to suggest it is the old version still running.
> 
> ```
> (running) since Tue 2023-05-02 20:39:32 UTC
> ```

Resolved conflicts. Also re-tested and confirmed it works as expected:

1. Install v1.13.0 using install script, verify service is running (because install script starts the service)
2. Install snapshot build from this branch using `yum install <path to rpm>`, notice that the service is **no longer running**
3. Manually start service
4. Make a commit, build new snapshot package, install package, notice that the service **remains running** due to the install script fix

At this point, the service is still running the previous install, because we have not manually restarted the service. This is consistent with RPM package conversions. The issue we are resolving here is that the service is stopped unexpectedly, when it should be left alone during package upgrade.

We can comfirm this with `bindplane version`
```bash
[jsirianni@rhel-test ~]$ sudo /usr/local/bin/bindplane --config /etc/bindplane/config.yaml version

client: v1.14.0-SNAPSHOT-50d8c366 (50d8c3666b4f9398df3ffe411c0b4ca8c3d806a1)
server: v1.14.0-SNAPSHOT-120873f4 (120873f4a0f61ad7826b91a9bba1cfed712e0fd1)
```

After manually restarting

```bash
[jsirianni@rhel-test ~]$ sudo /usr/local/bin/bindplane --config /etc/bindplane/config.yaml version
client: v1.14.0-SNAPSHOT-50d8c366 (50d8c3666b4f9398df3ffe411c0b4ca8c3d806a1)
server: v1.14.0-SNAPSHOT-50d8c366 (50d8c3666b4f9398df3ffe411c0b4ca8c3d806a1)
```

NOTE: When using the install script to perform an upgrade, service restart is called automatically for the user. This PR addresses an issue where the package install / upgrade / remove scripts are handling upgrade improperly, and stopping the service instead of leaving it be.

File: test/inspec/package/remove/provision.sh [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+        sudo bash /tmp/data/install-linux.sh --file /tmp/data/bindplane_*_linux_amd64.deb
+    elif command -v "rpm" > /dev/null ; then
+        sudo bash /tmp/data/install-linux.sh --file /tmp/data/bindplane_*_linux_amd64.rpm
+    else
+        echo "failed to detect platform type"
+        exit 1
+    fi
+}
+
+# Modify the config to prevent package manager from revmoving it.
```
Comment (Mrod1598) [inline test/inspec/package/remove/provision.sh:29]: remove is misspelled

**Review by Mrod1598 — APPROVED** [review]

**Pull request merged**